File: backend/app/models/pin.rb
# app/models/pin.rb
class Pin
  include Mongoid::Document
  include Mongoid::Timestamps

  field :user_id, type: BSON::ObjectId
  field :collection_id, type: BSON::ObjectId
  field :pin_type, type: String # 'restaurant', 'concert', 'sports'
  field :title, type: String
  field :summary, type: String
  field :ai_recommendation, type: String # from Snowflake
  field :location, type: Hash # { address, lat, lng, place_id }
  field :datetime, type: DateTime # for events
  field :source_images, type: Array, default: [] # URLs from Uploadthing
  field :links, type: Hash, default: {} # { website, tickets, menu, reviews }
  field :metadata, type: Hash, default: {} # flexible for different pin types
  field :processing_status, type: String, default: 'pending' # 'pending', 'processing', 'complete', 'failed'

  belongs_to :user
  belongs_to :pin_collection, class_name: 'Collection', optional: true, foreign_key: 'collection_id'

  validates :user_id, presence: true
  validates :processing_status, inclusion: { in: %w[pending processing complete failed] }
  validates :pin_type, inclusion: { in: %w[restaurant concert sports event] }, allow_nil: true

  index({ user_id: 1, created_at: -1 })
  index({ processing_status: 1 })
  index({ collection_id: 1 })

  scope :completed, -> { where(processing_status: 'complete') }
  scope :for_user, ->(user_id) { where(user_id: user_id) }
  scope :in_collection, ->(collection_id) { where(collection_id: collection_id) }
end

File: frontend/src/lib/types/pin.ts
import { z } from 'zod';

export const ProcessingStatusSchema = z.enum(['pending', 'processing', 'complete', 'failed']);

export const LocationSchema = z.object({
  address: z.string().optional(),
  lat: z.number().optional(),
  lng: z.number().optional(),
  place_id: z.string().optional(),
  google_maps_url: z.string().url().optional(),
});

export const LinksSchema = z.object({
  website: z.string().url().optional(),
  tickets: z.string().url().optional(),
  menu: z.string().url().optional(),
  reviews: z.string().url().optional(),
});

export const PinTypeSchema = z.enum(['restaurant', 'concert', 'sports', 'other']);

export const PinSchema = z.object({
  _id: z.string(),
  user_id: z.string(),
  collection_id: z.string().optional().nullable(),
  pin_type: PinTypeSchema.nullable(),
  title: z.string().nullable(),
  summary: z.string().nullable(),
  ai_recommendation: z.string().optional().nullable(),
  location: LocationSchema.optional().nullable(),
  datetime: z.string().datetime().optional().nullable(),
  source_images: z.array(z.string()),
  links: LinksSchema.optional().nullable(),
  metadata: z.record(z.unknown()).optional().nullable(),
  processing_status: ProcessingStatusSchema,
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
}).transform((data) => ({
  id: data._id,
  user_id: data.user_id,
  collection_id: data.collection_id,
  pin_type: data.pin_type ?? 'other',
  title: data.title ?? 'Untitled',
  summary: data.summary ?? '',
  ai_recommendation: data.ai_recommendation,
  location: data.location,
  datetime: data.datetime,
  source_images: data.source_images,
  links: data.links,
  metadata: data.metadata,
  processing_status: data.processing_status,
  created_at: data.created_at,
  updated_at: data.updated_at,
}));

export type ProcessingStatus = z.infer<typeof ProcessingStatusSchema>;
export type Location = z.infer<typeof LocationSchema>;
export type Links = z.infer<typeof LinksSchema>;
export type PinType = z.infer<typeof PinTypeSchema>;
export type Pin = z.infer<typeof PinSchema>;

export interface PinResponse {
  _id: string;
  user_id: string;
  collection_id?: string | null;
  pin_type: PinType | null;
  title: string | null;
  summary: string | null;
  ai_recommendation?: string | null;
  location?: Location | null;
  datetime?: string | null;
  source_images: string[];
  links?: Links | null;
  metadata?: Record<string, unknown> | null;
  processing_status: ProcessingStatus;
  created_at: string;
  updated_at: string;
}

export const CreatePinRequestSchema = z.object({
  image_url: z.string().url(),
  collection_id: z.string().optional(),
});

export const UpdatePinRequestSchema = z.object({
  collection_id: z.string().optional(),
  title: z.string().optional(),
  summary: z.string().optional(),
});

export type CreatePinRequest = z.infer<typeof CreatePinRequestSchema>;
export type UpdatePinRequest = z.infer<typeof UpdatePinRequestSchema>;
